src/zenml/config/store_config.py

```diff
@@ -0,0 +1,50 @@
+#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Functionality to support ZenML store configurations."""
+
+from pydantic import BaseModel
+
+from zenml.enums import StoreType
+from zenml.logger import get_logger
+
+logger = get_logger(__name__)
+
+
+class StoreConfiguration(BaseModel):
+    """Generic store configuration.
+
+    The store configurations of concrete store implementations must inherit from
+    this class and validate any extra attributes that are configured in addition
+    to those defined in this class.
+
+    Attributes:
+        type: The type of store backend.
+        url: The URL of the store backend.
+    """
+
+    type: StoreType
+    url: str
+
+    class Config:
+        """Pydantic configuration class."""
+
+        # Validate attributes when assigning them. We need to set this in order
+        # to have a mix of mutable and immutable attributes
+        validate_assignment = True
+        # Allow extra attributes to be set in the base class. The concrete
+        # classes are responsible for validating the attributes.
+        extra = "allow"
```

Comment:
Any specific reason to have this here? I'm guessing at some point the config get instantiated as a `StoreConfiguration` instead of the correct subclass?

Reply:
Ah, yes, you're correct. This allows me to reference `StoreConfiguration` in `GlobalConfiguration`, and I can also assign a `StoreConfiguration` object to a `SqlZenStoreConfiguration` field and let the `SqlZenStoreConfiguration` class deal with the validation.

Reply:
I ran into the exact same issue in the thing I'm currently working on. I found a different approach that makes sense in my situation, I'll just leave it here for you to decide if it's worth in this case as well:
```python
class StoreConfig(BaseModel):
    class Config:
        extra = "forbid"

class Container(BaseModel):
    store: StoreConfig
    
    @validator("config", pre=True)
    def _convert_config(cls, config: Any) -> StoreConfig:
        if isinstance(config, Dict):
            # this happens when it gets read from disk
            store_type = config["type"]
            ...  # create the right object here
```

Reply:
That's exactly what I was using, a validator to convert the base `StoreConfig` into the proper subclass, but what I found is that it's not even needed, because Pydantic does the conversion automatically. To expand on your example: 

```python
from pydantic import BaseModel

class StoreConfig(BaseModel):
    name: str

    class Config:
        extra = "allow"
        validate_assignment = True

class Container(BaseModel):
    store: StoreConfig

    class Config:
        extra = "forbid"
        validate_assignment = True

class CustomStoreConfig(StoreConfig):
    custom_attr1: str
    custom_attr2: int

    class Config:
        extra = "forbid"

class CustomContainer(Container):
    store: CustomStoreConfig

store = StoreConfig(name="test", custom_attr1="value", custom_attr2="1")
cont = CustomContainer(store=store)
```

Reply:
Ah I see, in my case the `CustomContainer` class doesn't exist and instead the `Container` can hold potentially different "stores", which is why the automatic conversion doesn't work